### PR TITLE
Cast data before saving it into database and mongoId as Relationships

### PR DESCRIPTION
**Pull request by RTLer**
in this PR i modify the eloquent cast part to cast data before saving it into DB. 
in project its difficult to cast data properly before saving it 
i add $saveCasts into model that can be filled  just $cast
mainly it use $saveCasts to cast Relationships into mongoId.
the mongoId as Relationships part can be activated by setting "use_mongo_id": 

``` php
        'mongodb' => [
            'name'         => 'mongodb',
            'driver'       => 'mongodb',
            'host'         => '127.0.0.1',
            'database'     => 'unittest',
            'use_mongo_id' => true,
        ]
```

also this PR contains new test's for use_mongo_id part and some fixes to make it work with mongoIds

### CODE REVIEW

**Comment by amiri27** [comment]
+1 it.. Nice job guy..

**Comment by RTLer** [comment]
see #834

**Comment by OSDDQD** [comment]
Can u add this in master please?

**Comment by Eke** [comment]
+1

**Comment by Eke** [comment]
Tested this one, at this moment works great for me. @RTLer Thanks !

**Comment by amiri27** [comment]
@jenssegers ignoring community pull requests is not an acceptable behavior..

**Comment by jenssegers** [comment]
@amiri27 It's on my todo list. There's quite some code to review.

**Comment by amiri27** [comment]
Well done @jenssegers ... Thanks a lot for your great library..

**Comment by demotivationme** [comment]
Some news? When this one be released? Cause I don't really want to change ObjectID relations to string

**Comment by amiri27** [comment]
@jenssegers still no update after 1 month..

**Comment by hotrush** [comment]
@jenssegers why not merged?

**Comment by deividaspetraitis** [comment]
@jenssegers any updates according this?

**Comment by sharmasabhay** [comment]
@jenssegers do you have this update in your latest release

**Comment by pi0** [comment]
There are breaking changes on master so unitTests are failing. would you please follow [This](https://github.com/jenssegers/laravel-mongodb/issues/937) steps to rebase this PR with master branch and running UnitTests against latest changes ?

Thanks :)

**Comment by RTLer** [comment]
thx @pi0, i'm update the branch with the latest changes on master and now all tests pass again

File: patch.txt [hunk 1]
```diff
@@ -0,0 +1,46 @@
+diff --git a/src/Jenssegers/Mongodb/Eloquent/Model.php b/src/Jenssegers/Mongodb/Eloquent/Model.php
```
Comment (pi0) [inline patch.txt]: Should not this diff patch file be excluded from commit ?

File: src/Jenssegers/Mongodb/Eloquent/Model.php [hunk 2]
```diff
@@ -552,4 +547,163 @@ public function __call($method, $parameters)
...
+     * @param $key
+     */
+    public function setRelationCast($key)
+    {
+        if ($key == '_id') {
+            $this->saveCasts['_id'] = 'ObjectID';
+            return;
+        }
+        if ($this->useMongoId()) {
+            if (ends_with($key, '_id')) {
```
Comment (pi0) [inline src/Jenssegers/Mongodb/Eloquent/Model.php]: What about relations ending with `_ids` ?

File: tests/config/database.php [hunk 3]
```diff
@@ -5,10 +5,11 @@
...
         'mongodb' => [
-            'name'       => 'mongodb',
-            'driver'     => 'mongodb',
-            'host'       => '127.0.0.1',
-            'database'   => 'unittest',
+            'name'         => 'mongodb',
+            'driver'       => 'mongodb',
+            'host'         => '127.0.0.1',
+            'database'     => 'unittest',
+            'use_mongo_id' => true,
```
Comment (pi0) [inline tests/config/database.php]: we where already enabling `use_mongo_id` in `RelationsWithMongoIdTest`  `setUp` method ! why enforcing it here again? (if this tast runs after other tests then `tearDown` will be called later and every other test before that will be using mongo_ids enabled !

**Comment by pi0** [comment]
+1 for merge!
@RTLer it seems some diff file is accidentally committed but thanks for this nice patch :)
@jenssegers is there any chance make a review/merge tonight? We already merged this PR to our fork and it was really working fine .. :)

**Comment by amiri27** [comment]
@jenssegers can you merge these changes?

**Comment by pi0** [comment]
Some changes may required before merge !?

**Comment by amiri27** [comment]
@jenssegers guy could you please respect to community by reply to them? still waiting to see any changes regarding this PR?

**Comment by RTLer** [comment]
it's been long time since i opened this PR, i think it's not going to merge, so i'm closing this PR.

if anyone wants this functionality it can be found on ["moloquent/moloquent"](https://moloquent.github.io/master/) (it's based on jenssegers/laravel-mongodb but active!)

**Comment by umimehar** [comment]
@jenssegers Any update, I want to use ObjectIds as relations identifier as I am sharing database between multiple servers with multiple programming languages. having a problem that when we insert a document in mongo using this lib, it stores object id of other collection as a string. and other drivers in other languages uses ObjectIds to populate the document from relation. 
Any news about this update?

**Comment by Smolevich** [comment]
@umimehar, see https://github.com/jenssegers/laravel-mongodb/issues/1892

**Comment by rennokki** [comment]
@umimehar Open this PR on https://github.com/rennokki/laravel-mongodb

**Pull request closed without merging**